matrix/lambdas/api/v1/core.py

```diff
@@ -0,0 +1,266 @@
+    # if the request tracker is not able to retrieve the format,
+    # it means that the driver has not created the relevant entry in the output table yet.
+    try:
+        format = request_tracker.format
+    except MatrixException:
+        return in_progress_response
+
+    # Failed case
+    if request_tracker.error:
+        return ({'request_id': request_id,
+                 'status': MatrixRequestStatus.FAILED.value,
+                 'matrix_url': "",
+                 'eta': "",
+                 'message': request_tracker.error},
+                requests.codes.ok)
+    # Check for failed batch conversion job
+    elif request_tracker.batch_job_status and request_tracker.batch_job_status == "FAILED":
+        request_tracker.log_error("The matrix conversion as a part of the request has failed. \
+            Please retry or contact an hca admin for help.")
+        return ({'request_id': request_id,
+                 'status': MatrixRequestStatus.FAILED.value,
+                 'matrix_url': "",
+                 'eta': "",
+                 'message': request_tracker.error},
+                requests.codes.ok)
+
+    # Complete case
+    elif request_tracker.is_request_complete():
+        matrix_results_bucket = os.environ['MATRIX_RESULTS_BUCKET']
+
+        matrix_location = ""
+        if format == MatrixFormat.LOOM.value:
+            matrix_location = f"https://s3.amazonaws.com/{matrix_results_bucket}/{request_id}.{format}"
+        elif format == MatrixFormat.CSV.value or format == MatrixFormat.MTX.value:
+            matrix_location = f"https://s3.amazonaws.com/{matrix_results_bucket}/{request_id}.{format}.zip"
+
+        return ({'request_id': request_id,
+                 'status': MatrixRequestStatus.COMPLETE.value,
+                 'matrix_url': matrix_location,
+                 'eta': "",
+                 'message': f"Request {request_id} has successfully completed. "
+                            f"The resultant expression matrix is available for download at "
+                            f"{matrix_location}"},
+                requests.codes.ok)
+
+    # Timeout case
+    elif request_tracker.timeout:
+
+        return ({'request_id': request_id,
+                 'status': MatrixRequestStatus.FAILED.value,
+                 'matrix_url': "",
+                 'eta': "",
+                 'message': request_tracker.error},
+                requests.codes.ok)
+    else:
+        return in_progress_response
+
+
+def get_filters():
+    """Return the list of accepted filter fields, like
+        ["project.short_name", "library_preparation.strand", ...
+    """
+    return (list(constants.FILTER_DETAIL.keys()),
+            requests.codes.ok)
+
+
+def get_fields():
+    """Return the list of accepted fields, like
+        ["project.short_name", "library_preparation.strand", ...
+    """
+    return (list(constants.FIELD_DETAIL.keys()),
+            requests.codes.ok)
+
+
+def _redshift_detail_lookup(name, description):
```

Comment:
In order to reduce the chance of API gateway timeouts, could `FIELD_DETAIL_..._QUERY_TEMPLATE`s accept `table_name` and use a single LEFT OUTER JOIN (`PRIMARY_KEY` mapping exists in `RedshiftHandler`)?

Reply:
Yeah I think that should work.